playpen/repo_level_awareness/agent/dependency_agent/dependency_agent.py

```diff
@@ -0,0 +1,347 @@
+                    find_pom_lines = m(a.code)
+
+        # We are going to give back the response, The caller should be responsible for running the code generated by the AI.
+        # If we have not take the actions step wise, in the LLM, we need to run all but editor here
+        # and give that information to the caller.
+        return MavenDependencyResult(
+            encountered_errors=None,
+            modified_files=None,
+            final_answer=llm_response.final_answer,
+            fqdn_response=maven_search,
+            find_in_pom=find_pom_lines,
+        )
+
+    def parse_llm_response(
+        self, content: str | List[str] | Dict
+    ) -> Optional[_llm_response]:
+        # We should not expect that the value is anything other than str for the type of
+        # call that we know we are making
+        if isinstance(content, Dict) or isinstance(content, List):
+            return None
+
+        actions = []
+        in_code = False
+        in_final_answer = False
+        in_thought = False
+        in_observation = False
+        code_block = ""
+        thought_str = ""
+        observation_str = ""
+        final_answer = ""
+        for line in content.splitlines():
+            if (
+                not line.strip().strip("```")  # trunk-ignore(ruff/B005)
+                or line == "```python"
+            ):
+                continue
+
+            parts = line.split(":")
+
+            if len(parts) > 1:
+                match parts[0]:
+                    case "Thought":
+                        s = ":".join(parts[1:])
+                        if code_block or observation_str:
+                            actions.append(
+                                _action(code_block, thought_str, observation_str)
+                            )
+                            code_block = ""
+                            thought_str = ""
+                            observation_str = ""
+                            in_observation = False
+                        thought_str = s.strip()
+                        in_thought = True
+                        continue
+                    case "Action":
+                        in_code = True
+                        in_thought = False
+                        continue
+                    case "Observation":
+                        s = ":".join(parts[1:])
+                        observation_str = s.strip()
+                        in_code = False
+                        in_observation = True
+                        continue
+                    case "Final Answer":
+                        actions.append(
+                            _action(code_block, thought_str, observation_str)
+                        )
+                        in_final_answer = True
+                        in_code = False
+                        in_thought = False
+                        continue
+
+            # TODO: There has to be a better way with python to do this.
```

Comment:
something like

```python
# Initialize the sections with their respective flags and configurations
sections = {
    'code_block': {'lines': [], 'strip_line': True, 'active_flag': in_code},
    'final_answer': {'lines': [], 'strip_line': False, 'active_flag': in_final_answer},
    'thought_str': {'lines': [], 'strip_line': True, 'active_flag': in_thought},
    'observation_str': {'lines': [], 'strip_line': False, 'active_flag': in_observation},
}

# Process each line and append to the appropriate section
for line in lines:
    for section in sections.values():
        if section['active_flag']:
            line_content = line.strip() if section['strip_line'] else line
            section['lines'].append(line_content)

# Construct the final content for each section
code_block = '\n'.join(sections['code_block']['lines']).strip()
final_answer = '\n'.join(sections['final_answer']['lines']).strip()
thought_str = '\n'.join(sections['thought_str']['lines']).strip()
observation_str = '\n'.join(sections['observation_str']['lines']).strip()
```

could work and do the same?

Reply:
obviously you'd initialize outside the loop and just set the flags manually

Reply:
I don't think this is necessarily better though

Reply:
I think that we are going to leave it, and come back to see if that makes sense?

Going to resolve go ahead and re-open it if you disagree